comment/tests/test_mixins.py

```diff
@@ -16,131 +21,175 @@ def test_non_ajax_request(self):
-        request = self.factory.get(self.get_url(**url_data))
-        response = self.mixin.dispatch(request)
+    def test_edit_comment_by_anonymous(self):
+        """anonymous will be redirected to login page"""
+        self.client.logout()
+        response = self.client.post(self.url, data=self.data, HTTP_X_REQUESTED_WITH='XMLHttpRequest')
+        self.assertEqual(response.status_code, 302)
+        self.assertEqual(response.url, settings.LOGIN_URL + f'?next=/comment/edit/{self.comment.id}/')
 
-        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
-        self.assertEqual(self.mixin.error, f'{model_id} is NOT a valid model id for the model {url_data["model_name"]}')
 
-    def test_model_id_non_integral(self):
-        url_data = self.url_data.copy()
-        model_id = 'not integral'
-        url_data['model_id'] = model_id
-        request = self.factory.get(self.get_url(**url_data))
-        response = self.mixin.dispatch(request)
+class CanDeleteMixinTest(BaseCommentMixinTest):
+    def setUp(self):
+        super().setUp()
+        self.client.force_login(self.user_1)
+        self.comment = self.create_comment(self.post_1, self.user_1)
+        self.url = reverse('comment:delete', kwargs={'pk': self.comment.id})
+
+    def test_delete_comment_by_owner(self):
+        self.assertEqual(self.comment.user.id, self.user_1.id)
+        response = self.client.post(self.url, data=self.data, HTTP_X_REQUESTED_WITH='XMLHttpRequest')
+        self.assertEqual(response.status_code, 200)
+
+    def test_delete_comment_by_non_owner(self):
+        self.assertNotEqual(self.comment.user.id, self.user_2.id)
+        self.client.force_login(self.user_2)
+        response = self.client.post(self.url, data=self.data, HTTP_X_REQUESTED_WITH='XMLHttpRequest')
+        self.assertEqual(response.status_code, 403)
+
+    def test_delete_comment_by_admin(self):
+        self.assertNotEqual(self.comment.user.id, self.admin.id)
+        self.client.force_login(self.admin)
+        response = self.client.post(self.url, data=self.data, HTTP_X_REQUESTED_WITH='XMLHttpRequest')
+        self.assertEqual(response.status_code, 200)
+
+    def test_delete_comment_by_moderator(self):
+        """moderator cannot delete comment unless it's flagged"""
+        self.assertNotEqual(self.comment.user.id, self.moderator.id)
+        self.client.force_login(self.moderator)
+        response = self.client.post(self.url, data=self.data, HTTP_X_REQUESTED_WITH='XMLHttpRequest')
+        self.assertEqual(response.status_code, 403)
+
+    def test_delete_flagged_comment_by_moderator(self):
+        """moderator cannot delete comment unless it's flagged"""
+        self.assertNotEqual(self.comment.user.id, self.moderator.id)
+        self.client.force_login(self.moderator)
+        self.create_flag_instance(self.user_2, self.comment)
+        self.create_flag_instance(self.admin, self.comment)
+        self.create_flag_instance(self.moderator, self.comment)
```

Comment:
`create_flag_instance` won't work here since it is a class method now. Please use `set_flag`.

Reply:
What do you mean? I actually used it and the comment indeed became flagged. BTW, it took me so long debugging this test case and finally found a silly bug in changing the flag state method https://github.com/Radi85/Comment/pull/67/files#diff-76e68827e2becc6ab78bf6afd4659de9L90
The state was changing to unflagged after flagging a flagged comment 😄

Reply:
`create_flag_instance` is a class method?

> The state was changing to unflagged after flagging a flagged comment smile

why would that happen?  if I'm in my most stable state, self.state == self.FLAGGED would not be true !!!

Reply:
```python
if self.count > allowed_flags and self.state == self.UNFLAGGED:
```
`self.state == self.UNFLAGGED` this will be false when the comment is already flagged, therefore it will hit the else statment and unflag the comment though `self.count = allowed_flags + 2`

Reply:
i am not sure i completely understand currently why an else statement is necessary there. i will re-check when i get up tomorrow.

Reply:
yes, the else part will be useful when people delete their flags.

Reply:
but why is `self.create_flag_instance` working here?

Reply:
why should not?

Reply:
it is a class method, you're calling it by passing an instance.

Reply:
As I know, any class property can be accessed by the instance.